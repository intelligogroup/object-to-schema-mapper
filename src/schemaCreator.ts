import { SomeObj } from './utils/types';
import { resolveValueArchtype, resolveType } from './utils/objectUtil';

function mapObjectToSchema(obj: SomeObj, schema: SomeObj = {}): SomeObj {
    return Object
        .keys(obj)
        .reduce(
            (schema, key) => resolveOperation(schema, obj, key),
            schema
        );
}

function handleEmptyValue(schema: SomeObj, key: string) {
    const value = schema[key];
    if (value == null || value == 'unknown') {
        return unknownType(schema, key);
    }
    return schema;
}

function resolveOperation(schema: SomeObj, obj: any, key: string): SomeObj {
    const nestedValue = obj[key];
    const archType = resolveValueArchtype(nestedValue);
    switch (archType) {
        case 'EMPTY':
            return handleEmptyValue(schema, key);
        case 'SIMPLE':
            const type = resolveType(nestedValue);
            return simpleType(type, key, schema);
        case 'ARRAY':
            return Object.assign(schema, { [key]: [arrayOfObjectsToSchema(nestedValue, {})] });
        case 'OBJECT':
            return Object.assign(schema, { [key]: mapObjectToSchema(nestedValue, {}) });
    }
}

function arrayOfObjectsToSchema(array: [any], schema: SomeObj = {}): SomeObj {
    return array.reduce(
        (acc, obj) => resolveValueArchtype(obj) == 'OBJECT'
            ? mapObjectToSchema(obj, acc)
            : JSON.stringify({ type: resolveType(obj).toLowerCase() }),
        schema
    );
}

function simpleType(type: string, key: string, schema: SomeObj) {
    return Object.assign(schema, { [key]: JSON.stringify({ type: type.toLowerCase() }) });
}

function unknownType(schema: SomeObj, key: string) {
    return Object.assign(schema, { [key]: JSON.stringify({ type: 'unknown' }) });
}

export { mapObjectToSchema };
