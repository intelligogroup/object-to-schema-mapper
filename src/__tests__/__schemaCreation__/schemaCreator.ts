import data from './data.json';
import { mapObjectToSchema } from '../../schemaCreator';

describe('map object to schema tests', () => {

    test('mapper creates a schema that encompases all data from multiple results', () => {
        const createdSchema = data.info.reduce(
            (accSchema, object) => mapObjectToSchema(object, accSchema),
            {}
        )
        expect(createdSchema).toMatchSnapshot()
    })

})
