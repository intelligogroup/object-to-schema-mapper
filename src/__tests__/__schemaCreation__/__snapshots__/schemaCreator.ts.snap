// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`map object to schema tests mapper creates a schema that encompases all data from multiple results 1`] = `
Object {
  "person": Object {
    "addes": Array [
      Object {
        "house": "{\\"type\\":\\"string\\",\\"example\\":\\"h2\\"}",
        "street": "{\\"type\\":\\"string\\",\\"example\\":\\"s2\\"}",
      },
    ],
    "addesB": Array [
      Object {
        "house": "{\\"type\\":\\"string\\",\\"example\\":\\"h2B\\"}",
        "street": "{\\"type\\":\\"string\\",\\"example\\":\\"s2B\\"}",
      },
    ],
    "currentPosition": "{\\"type\\":\\"string\\",\\"example\\":\\"e\\"}",
    "dateOfBirth": "{\\"type\\":\\"string\\",\\"example\\":\\"10/14/88\\"}",
    "interests": Array [
      "{\\"type\\":\\"array\\",\\"example\\":\\"alcoholism\\"}",
    ],
    "name": Object {
      "mname": "{\\"type\\":\\"string\\",\\"example\\":\\"MNAME\\"}",
    },
    "pastPositions": Array [
      "{\\"type\\":\\"array\\",\\"example\\":\\"a\\"}",
    ],
    "taho": Array [
      Object {
        "id": "{\\"type\\":\\"number\\",\\"example\\":3}",
        "nd": Object {
          "test": "{\\"type\\":\\"number\\",\\"example\\":53}",
        },
        "test": Array [
          Object {
            "a": "{\\"type\\":\\"number\\",\\"example\\":6}",
            "foo": "{\\"type\\":\\"string\\",\\"example\\":\\"f6\\"}",
          },
        ],
        "test0": "{\\"type\\":\\"number\\",\\"example\\":11}",
        "test1": "{\\"type\\":\\"number\\",\\"example\\":22}",
        "test2": "{\\"type\\":\\"number\\",\\"example\\":33}",
      },
    ],
  },
}
`;
